lru/lru.go

```diff
@@ -0,0 +1,101 @@
+package lru
+
+import "container/list"
+import "sync"
+
+// Cache represents a generic LRU cache with a very simple interface.
+type Cache interface {
+	// Put stores the given (key,value) pair, replacing existing value if key
+	// already exists.
+	Put(key, value interface{}) error
+
+	// Get returns the value for a given key.
+	Get(key interface{}) (interface{}, error)
+
+	// Len returns number of elements in the cache.
+	Len() int
+}
+
+// elementMap is an alias for a map from a generic interface to a list.Element.
+type elementMap map[interface{}]*list.Element
+
+// entry represents a (key,value) pair entry in the Cache. The Cache's list
+// stores entries which let us get the cache key when an entry is evicted.
+type entry struct {
+	key   interface{}
+	value interface{}
+}
+
+// LRUCache provides a generic thread-safe lru cache that can be used for
+// storing filters, blocks, etc.
+type LRUCache struct {
+	// size is the capacity of the cache (ie not the number of elements in it).
+	size int
+
+	// ll is a doubly linked list which keeps track of recency of used elements
+	// by moving them to the front.
+	ll *list.List
+
+	// cache is a generic cache which allows us to find an elements position
+	// in the ll list from a given key.
+	cache elementMap
+
+	// mtx is used to make sure the LRUCache is thread-safe.
+	mtx sync.RWMutex
+}
+
+// NewLRUCache return a cache with specified size, it can't contain more than
+// that many elements.
+func NewLRUCache(size int) *LRUCache {
+	return &LRUCache{
+		size:  size,
+		ll:    list.New(),
+		cache: make(elementMap),
+	}
+}
+
+// Put inserts a given (key,value) pair into the cache, if the key already
+// exists, it will replace value and update it to be most recent item in cache.
+func (c *LRUCache) Put(key, value interface{}) error {
+	c.mtx.Lock()
+	defer c.mtx.Unlock()
+
+	// key already exists, replace value.
+	if el, ok := c.cache[key]; ok {
+		c.ll.MoveToFront(el)
+		el.Value.(*entry).value = value
+		return nil
+	}
+
+	el := c.ll.PushFront(&entry{key, value})
+	c.cache[key] = el
+
+	if c.ll.Len() > c.size {
+		if elr := c.ll.Back(); elr != nil {
+			c.ll.Remove(elr)
+			delete(c.cache, elr.Value.(*entry).key)
+		}
+	}
+
+	return nil
+}
+
+// Get will return value for a given key, making the element the most recently
+// accessed item in the process. Will return nil if the key isn't found.
+func (c *LRUCache) Get(key interface{}) (interface{}, error) {
+	c.mtx.RLock()
+	defer c.mtx.RUnlock()
+
+	if el, ok := c.cache[key]; ok {
+		c.ll.MoveToFront(el)
```

Comment:
suggestion: make happy-path the last returned, i.e. check `!ok` here and return nil in that case (and include a comment) :)